### PR TITLE
add sparse gradients support for RMSPropOptimizier

### PR DESCRIPTION
**Pull request by suiyuan2009**
#2476

### CODE REVIEW

File: tensorflow/core/ops/ops.pbtxt [hunk 1]
```diff
@@ -11111,6 +11111,104 @@ op {
   description: "That is for rows we have grad for, we update var and accum as follows:\n\naccum = accum * momentum + grad\nvar -= lr * accum"
 }
 op {
+  name: "SparseApplyRMSProp"
```
Comment (vrv) [inline tensorflow/core/ops/ops.pbtxt]: you can drop this change -- we automatically update the ops now, I believe

File: tensorflow/python/training/rmsprop_test.py [hunk 2]
```diff
@@ -23,9 +23,132 @@
 import numpy as np
 import tensorflow as tf
 
+def rmsprop_update_numpy(var, g, rms, mom, lr, decay, momentum, epsilon):
```
Comment (vrv) [inline tensorflow/python/training/rmsprop_test.py]: you can define this within the class RMSPropOptimizerTest

**Comment by lukaszkaiser** [comment]
George Dahl agreed to take a look, he's just joining the right groups on github.

**Comment by georgedahl** [comment]
I took a very quick glance and I believe this PR does the "wrong" thing efficiently, however it is a pretty reasonable "wrong" thing to do. What I mean is that for sparse updates, I don't think this will behave the same was as treating them as dense would. So doing the same optimization, but treating the updates as sparse will produce different results than treating them as dense, even if all the gradients are the same. This property needs to be documented, because one might somehow assume that using the sparse version would do the same thing as the dense one, just more efficiently when you have sparse updates.

Essentially, RMSProp and similar algorithms generate an update for each weight at each step and the approach here discards the updates for weights that have a zero gradient on that step. This gives the rarely updated indices a lower learning rate than they would otherwise have in the dense version of the algorithm since "correct" RMSprop would divide by gradient statistics and amplify the updates for weights that had many zero gradients. However, that said, for sparse updates this might work ok sometimes and it isn't clear what the best thing to do is, so doing the easy and efficient thing could be worth implementing.

I am pretty sure Adam in TF for sparse updates has the same problem as exhibited in this PR, and since this PR was for a feature request spawned by the inconsistency between ADAM supporting sparse updates and RMSProp not supporting them, it makes sense to have an analogous behavior. Of course it would be good to add something to the ADAM docs explaining that for sparse updates it only updates weights with non-zero gradient that step and makes no attempt at correcting for this.

So to summarize, the basic approach of the PR seems OK, but I would like to see the exact algorithm documented better in the sparse case so users can be warned not to think it works the same way as the dense update algorithm. ADAM might also need a doc improvement.

File: tensorflow/core/kernels/training_ops.cc [hunk 3]
```diff
@@ -1428,4 +1428,140 @@ REGISTER_KERNELS(GPU, double);
...
+      auto ms_flat = ms.flat_outer_dims<T>();
+      auto mom_flat = mom.flat_outer_dims<T>();
+      auto grad_flat = grad.flat_outer_dims<T>();
+      const T lr_scalar = lr.scalar<T>()();
+      const T rho_scalar = rho.scalar<T>()();
+      const T epsilon_scalar = epsilon.scalar<T>()();
+      const T momentum_scalar = momentum.scalar<T>()();
+
+      for (Tindex i = 0; i < N; i++) {
+	const Tindex index = indices_vec(i);
```
Comment (vrv) [inline tensorflow/core/kernels/training_ops.cc]: can you run clang-format (style google) on this file to fix a lot of the alignment issues in this file?

**Comment by suiyuan2009** [comment]
@georgedahl thank you very much! I took a look at 'RMSProp algorithm with Nesterov momentum' presented in Bengio's 'deep learning', you can record a variable like timestamp to calculate how many times you have not multiplied decay rate, but the simple implement is good enough maybe.

File: tensorflow/g3doc/api_docs/python/functions_and_classes/tf.train.AdamOptimizer.md [hunk 4]
```diff
@@ -33,6 +33,10 @@ The default value of 1e-8 for epsilon might not be a good default in
 general. For example, when training an Inception network on ImageNet a
 current good choice is 1.0 or 0.1.
 
+Note that in dense implement of this algorithm, m_t, v_t and variable will
```
Comment (vrv) [inline tensorflow/g3doc/api_docs/python/functions_and_classes/tf.train.AdamOptimizer.md]: you can drop the changes to the .md files, they are autogenerated

**Comment by vrv** [comment]
Sorry, you should comment when you've pushed new changes, otherwise we don't know when the code is ready.  You'll have to resolve conflicts and then update us when you've done so.

**Comment by suiyuan2009** [comment]
@vrv modification of .md has been removed and conflicts have been solved.
but merges too many commits and someone has not signed CLA

**Comment by vrv** [comment]
I think you rebased incorrectly, can you try again?

**Comment by suiyuan2009** [comment]
@vrv seems ok now

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by suiyuan2009** [comment]
@vrv MacOS CPU Tests failed, but Linux CPU Tests succeeded

```
//tensorflow/core:common_runtime_direct_session_with_tracking_alloc_test FAILED in 0.1s
  /private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/tensorflow-pull-requests-mac/bazel-out/local_darwin-opt/testlogs/tensorflow/core/common_runtime_direct_session_with_tracking_alloc_test/test.log

Executed 402 out of 517 tests: 516 tests pass and 1 fails locally.
```

**Comment by vrv** [comment]
I think it's a flaky test, i will try again  @tensorflow-jenkins test this please
